### ISSUE TITLE
Memory region APIs are unsound

### ISSUE THREAD

**Issue by DemiMarie** [cloud-hypervisor/cloud-hypervisor#6547]
**Describe the bug**
The memory region APIs are not marked `unsafe`, but they take arbitrary addresses (represented as `u64`).  The addresses are used so this is unsound.  Furthermore, the fields of `UserMemoryRegion` are all public, so one can create an invalid `UserMemoryRegion` in safe code.

**To Reproduce**
Call the memory region APIs with junk `userspace_addr` parameters.

**Version**
Latest git main.

**VM configuration**
Not applicable.  This is a problem with Cloud Hypervisor’s internal APIs.

**Logs**
Not applicable.  This is a problem with Cloud Hypervisor’s internal APIs.

**Comment by liuw** [comment]
Are you proposing to mark those APIs are unsafe? As you already noted, those are internal APIs, so the users are expected  to know what they are doing.

**Comment by DemiMarie** [comment]
Yes, they should be marked as `unsafe`, or wrapped in a safe abstraction.

**Comment by liuw** [comment]
Are you going to propose a patch? We probably only need to make `make_user_memory_region` unsafe and then hide all its fields.

**Comment by DemiMarie** [comment]
I suspect the hard part will be writing the `// SAFETY` comments at the call sites.  I’m not familiar enough with Cloud Hypervisor to do that.

**Comment by Pivnoy** [comment]
Hello!
Is this task still relevant?
And if the answer is "yes",can I  be assigned to contribute?

**Comment by liuw** [comment]
> Hello! Is this task still relevant? And if the answer is "yes",can I be assigned to contribute?

Yes. I've assigned this to you. Feel free to ask questions.
